cid.go

```diff
@@ -680,3 +680,129 @@ func CidFromBytes(data []byte) (int, Cid, error) {
 
 	return l, Cid{string(data[0:l])}, nil
 }
+
+func toBufByteReader(r io.Reader, dst []byte) *bufByteReader {
+	// If the reader already implements ByteReader, use it directly.
+	// Otherwise, use a fallback that does 1-byte Reads.
+	if br, ok := r.(io.ByteReader); ok {
+		return &bufByteReader{direct: br, dst: dst}
+	}
+	return &bufByteReader{fallback: r, dst: dst}
+}
+
+type bufByteReader struct {
+	direct   io.ByteReader
+	fallback io.Reader
+
+	dst []byte
+}
+
+func (r *bufByteReader) ReadByte() (byte, error) {
+	// The underlying reader has ReadByte; use it.
+	if br := r.direct; br != nil {
+		b, err := br.ReadByte()
+		if err != nil {
+			return 0, err
+		}
+		r.dst = append(r.dst, b)
+		return b, nil
+	}
+
+	// Fall back to a one-byte Read.
+	var p [1]byte
```

Comment:
I did think about using dst directly, but I didn't want to overcomplicate the code right away without benchmarks. It's a complete guess that it would be faster, I think, given that it's just one byte.

Are you sure that p allocates? We don't have "parse" benchmarks, so it's not easy to tell. From a quick check, I didn't see a "cid.go: p escapes to heap" from `go build -gcflags=-m` on Go 1.16.x.

Reply:
Fair enough to not over complicate now. `r.fallback` is an arbitrary reader so go can't know what `r.fallback.Read` does. So it needs to put `p` on the heap because `r.fallback.Read` could do anything with it.

It's probably not showing up now due to inlining? That's my guess. Given a small enough example, go may be able to determine that `fallback` can _only_ be one specific type of reader.

Reply:
I'm going to add this to "can be optimized once we have benchmarks" :) For now, this is easy, and escape analysis doesn't say it escapes when compiling the func, so... I think we're agreeing to leave as is.